crates/path_trie/src/lib.rs

```diff
@@ -0,0 +1,967 @@
+                .prefixes
+                .iter()
+                .max_by_key(|p| new_rank.get(p).copied().unwrap_or(usize::MAX));
+
+            let old_is_file = old_winner.is_some_and(|p| n.terminals.contains(p));
+            let new_is_file = new_winner.is_some_and(|p| n.terminals.contains(p));
+
+            match (old_is_file, new_is_file) {
+                (true, true) => {
+                    let old = old_winner.unwrap();
+                    let new = new_winner.unwrap();
+                    if old != new {
+                        to_clobbers.push((cur.clone(), old.clone()));
+                        from_clobbers.push((cur.clone(), new.clone()));
+                    }
+                    // don't descend
+                }
+                (false, true) => {
+                    let new = new_winner.unwrap();
+                    from_clobbers.push((cur.clone(), new.clone()));
+                    collect_removed_subtree(n, cur, old_rank, to_clobbers);
+                    // don't descend
+                }
+                (true, false) => {
+                    let old = old_winner.unwrap();
+                    to_clobbers.push((cur.clone(), old.clone()));
+                    collect_new_winners(n, cur, new_rank, from_clobbers);
+                    // don't descend further: new_winners handles it
+                }
+                (false, false) => {
+                    // Recurse normally
+                    for (comp, child) in &n.children {
+                        cur.push(comp);
+                        dfs(child, cur, old_rank, new_rank, to_clobbers, from_clobbers);
+                        cur.pop();
+                    }
+                }
+            }
+        }
+
+        {
+            let sorted_packages = self
+                .packages
+                .iter()
+                .cloned()
+                .sorted()
+                .collect::<Vec<PackageName>>();
+
+            let new_sorted_packages = new_order
+                .iter()
+                .cloned()
+                .sorted()
+                .collect::<Vec<PackageName>>();
+
+            assert_eq!(
+                &sorted_packages, &new_sorted_packages,
+                "Expected just reordering, got something else.
+Old:
+{sorted_packages:#?}
+New:
+{new_sorted_packages:#?}
+"
+            );
+        }
+
+        // Package with highest priority will have biggest number.
+        let old_rank = rank(self.packages.iter().rev());
+        let new_rank = rank(new_order.iter());
+
+        let mut to_clobbers = Vec::new();
+        let mut from_clobbers = Vec::new();
+
+        let mut buf = PathBuf::new();
+        dfs(
+            &self.root,
+            &mut buf,
+            &old_rank,
+            &new_rank,
+            &mut to_clobbers,
+            &mut from_clobbers,
+        );
+
+        self.packages.clear();
+        for pkg in new_order.iter().rev() {
+            self.packages.insert(pkg.clone());
+        }
```

Comment:
Lets avoid cloning and remove the for loop:

```suggestion
        self.packages.extend(new_order.into_iter().rev())
```